src/backend/access/transam/varsup.c

```diff
@@ -514,6 +520,76 @@ ForceTransactionIdLimitUpdate(void)
 	return false;
 }
 
+Oid
+GetNewTempObjectId(void)
+{
+	Oid			result;
+	Oid			tempOidStart;
+	static Oid 	nextTempOid = InvalidOid;
+
+	/* safety check, we should never get this far in a HS standby */
+	if (RecoveryInProgress())
+		elog(ERROR, "cannot assign OIDs during recovery");
+
+	if (!OidIsValid((Oid) temp_oid_buffer_start)) /* InvalidOid means it needs assigning */
+	{
+		/* First check to see if another connection has already picked a start, then update. */
+		LWLockAcquire(OidGenLock, LW_EXCLUSIVE);
+		if (OidIsValid(ShmemVariableCache->tempOidStart))
+		{
+			temp_oid_buffer_start = OID_TO_BUFFER_START(ShmemVariableCache->tempOidStart);
+			nextTempOid = ShmemVariableCache->tempOidStart;
+		}
+		else
+		{
+			/* We need to pick a new start for the buffer range. */
+			tempOidStart = ShmemVariableCache->nextOid;
+
+			/*
+			 * Decrement ShmemVariableCache->oidCount to take into account the new buffer we're allocating
+			 */
+			if (ShmemVariableCache->oidCount < temp_oid_buffer_size)
+				ShmemVariableCache->oidCount = 0;
+			else
+				ShmemVariableCache->oidCount -= temp_oid_buffer_size;
+
+			/*
+			 * If ShmemVariableCache->nextOid is below FirstNormalObjectId then we can start at FirstNormalObjectId here and
+			 * GetNewObjectId will return the right value on the next call.  
+			 */
+			if (tempOidStart < FirstNormalObjectId)
+				tempOidStart = FirstNormalObjectId;
+
+			/* If the OID range would wraparound, start from beginning instead. */
+			if (tempOidStart + temp_oid_buffer_size < tempOidStart) 
+				tempOidStart = FirstNormalObjectId;
+
+			temp_oid_buffer_start = OID_TO_BUFFER_START(tempOidStart);
+			ShmemVariableCache->tempOidStart = tempOidStart;
+			ShmemVariableCache->tempOidBufferSize = temp_oid_buffer_size;
+
+			nextTempOid = (Oid) tempOidStart;
+
+			/* Skip nextOid ahead to end of range here as well.  */
+			ShmemVariableCache->nextOid = (Oid) (tempOidStart + temp_oid_buffer_size);
```

Comment:
oidCount is decremented by the if statement above. In the scenario you're describing, oidCount will be set to 0. The next time GetNewObjectId is called, it will handle oidCount = 0. 

So the current behavior will have the buffer be a fixed 65k size - if there's fewer than that many OIDs actually available (say the last 10k OIDs in the buffer), then there is a chance we place the new OID buffer (since it will wraparound) in a zone of completely used OIDs and we may go into general OID wraparound behavior about 10k OIDs earlier than is expected. In my opinion this is an extreme edge case, since an application that would consume all the OIDs in the buffer likely won't be too affected by this case (which would only happen if the entire database consume the entire OID buffer before ever creating a temp table in Babelfish) - they would have to clean up OIDs eventually. 

The potential fixes for this edge case I think all raise questions of their own - if we limit the temp_oid_buffer_size to the size of the remaining ShmemVariableCache->oidCount, then we run the risk of having an extremely small buffer size with no way for the customer to fix it. We could disable the feature completely in the case where oidCount is less than the buffer size. However both of these solutions would still require some form of operator intervention.